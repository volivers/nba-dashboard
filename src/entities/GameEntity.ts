import { Game } from '../interfaces/Game';
import { TeamEntity, deserializeTeamEntity, mockedAwayTeamEntity, mockedHomeTeamEntity } from './TeamEntity';

export interface GameEntity {
  date: string;
  home_team: TeamEntity;
  home_team_score: number;
  id: number;
  period: number;
  postseason: boolean;
  season: number;
  status: string;
  time: string;
  visitor_team: TeamEntity;
  visitor_team_score: number;
}

export const deserializeGameEntity = (entity: GameEntity): Game => ({
  date: entity.date,
  homeTeam: deserializeTeamEntity(entity.home_team),
  homeTeamScore: entity.home_team_score,
  id: entity.id,
  period: entity.period,
  postseason: entity.postseason,
  season: entity.season,
  status: entity.status,
  time: entity.time,
  visitorTeam: deserializeTeamEntity(entity.visitor_team),
  visitorTeamScore: entity.visitor_team_score,
});

export const mockedGameEntity: GameEntity = {
  date: '2000-11-01T00:00:00.000Z',
  home_team: mockedHomeTeamEntity,
  home_team_score: 100,
  id: 1,
  period: 4,
  postseason: true,
  season: 1,
  status: 'Final',
  time: '',
  visitor_team: mockedAwayTeamEntity,
  visitor_team_score: 80,
};
