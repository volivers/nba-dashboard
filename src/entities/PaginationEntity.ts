import { Pagination } from '../interfaces/Pagination';

export interface PaginationEntity {
  current_page: number;
  next_page: number;
  per_page: number;
}

export const deserializePaginationEntity = (entity: PaginationEntity): Pagination => ({
  currentPage: entity.current_page,
  nextPage: entity.next_page,
  perPage: entity.per_page,
});

export const mockedPaginationEntity: PaginationEntity = {
  current_page: 1,
  next_page: 2,
  per_page: 25,
};
