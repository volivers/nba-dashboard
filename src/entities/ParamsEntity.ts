export interface ParamsEntity {
  page?: number;
  per_page?: number;
  game_ids?: number[];
  player_ids?: number[];
  team_ids?: number[];
}
