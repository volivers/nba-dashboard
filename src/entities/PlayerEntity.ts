import { Player } from '../interfaces/Player';
import { TeamEntity, deserializeTeamEntity, mockedHomeTeamEntity } from './TeamEntity';

export interface PlayerEntity {
  first_name: string;
  height_feet?: number;
  height_inches?: number;
  id: number;
  last_name: string;
  position: string;
  team: TeamEntity;
  weight_pounds?: number;
}

export const deserializePlayerEntity = (entity: PlayerEntity): Player => ({
  firstName: entity.first_name,
  heightFeet: entity?.height_feet,
  heightInches: entity?.height_inches,
  id: entity.id,
  lastName: entity.last_name,
  position: entity.position,
  team: deserializeTeamEntity(entity.team),
  weightPounds: entity?.weight_pounds,
});

export const mockedPlayerEntity: PlayerEntity = {
  first_name: 'Neemias',
  id: 88,
  last_name: 'Queta',
  position: 'C',
  team: mockedHomeTeamEntity,
};
