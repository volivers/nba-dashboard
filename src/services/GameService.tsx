import axios from 'axios';
import { PaginationEntity, deserializePaginationEntity } from '../entities/PaginationEntity';
import { Pagination } from '../interfaces/Pagination';
import { Params, serializeParams } from '../interfaces/Params';
import { GameEntity, deserializeGameEntity } from '../entities/GameEntity';
import { Game } from '../interfaces/Game';

export interface GameResponse {
  data: Game[];
  meta: Pagination;
}

export interface GameResponseEntity {
  data: GameEntity[];
  meta: PaginationEntity;
}

export const deserializeGameResponse = (entity: GameResponseEntity): GameResponse => ({
  data: entity.data.map(deserializeGameEntity),
  meta: deserializePaginationEntity(entity.meta)
});

export const GAMES_API_URL = process.env.REACT_APP_API_URL + '/games';

export class GameService {
  public static async getAll(params?: Params): Promise<GameResponse> {
    const response = await axios<GameResponseEntity>({
      method: 'GET',
      url: GAMES_API_URL,
      params: { ...serializeParams(params) },
      headers: {
        'Content-Type': 'application/json',
        'X-RapidAPI-Key': process.env.REACT_APP_API_KEY,
        'X-RapidAPI-Host': 'free-nba.p.rapidapi.com'
      }
    });

    return deserializeGameResponse(response.data);
  };
};
