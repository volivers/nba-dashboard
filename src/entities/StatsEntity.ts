import { Stats } from '../interfaces/Stats';
import { GameEntity, deserializeGameEntity, mockedGameEntity } from './GameEntity';
import { PlayerEntity, deserializePlayerEntity, mockedPlayerEntity } from './PlayerEntity';
import { TeamEntity, deserializeTeamEntity, mockedHomeTeamEntity } from './TeamEntity';

export interface StatsEntity {
  ast: number;
  blk: number;
  dreb: number;
  fg3_pct: number;
  fg3a: number;
  fg3m: number;
  fg_pct: number;
  fga: number;
  fgm: number;
  ft_pct: number;
  fta: number;
  ftm: number;
  game: GameEntity;
  id: number;
  min: string;
  oreb: number;
  pf: number;
  player: PlayerEntity;
  pts: number;
  reb: number;
  stl: number;
  team: TeamEntity;
  turnover: number;
}

export const deserializeStatsEntity = (entity: StatsEntity): Stats => ({
  ast: entity.ast,
  blk: entity.blk,
  dreb: entity.dreb,
  fg3Pct: entity.fg3_pct,
  fg3a: entity.fg3a,
  fg3m: entity.fg3m,
  fgPct: entity.fg_pct,
  fga: entity.fga,
  fgm: entity.fgm,
  ftPct: entity.ft_pct,
  fta: entity.fta,
  ftm: entity.ftm,
  game: deserializeGameEntity(entity.game),
  id: entity.id,
  min: entity.min,
  oreb: entity.oreb,
  pf: entity.pf,
  player: deserializePlayerEntity(entity.player),
  pts: entity.pts,
  reb: entity.reb,
  stl: entity.stl,
  team: deserializeTeamEntity(entity.team),
  turnover: entity.turnover,
});

export const mockedStatsEntity: StatsEntity = {
  ast: 10,
  blk: 4,
  dreb: 15,
  fg3_pct: 1,
  fg3a: 10,
  fg3m: 10,
  fg_pct: 10,
  fga: 10,
  fgm: 10,
  ft_pct: 10,
  fta: 10,
  ftm: 10,
  game: mockedGameEntity,
  id: 1,
  min: '18',
  oreb: 5,
  pf: 14,
  player: mockedPlayerEntity,
  pts: 20,
  reb: 10,
  stl: 5,
  team: mockedHomeTeamEntity,
  turnover: 10,
};
