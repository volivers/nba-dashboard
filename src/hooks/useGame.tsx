import { useQuery } from 'react-query';
import { GameService } from '../services/GameService';
import { Params } from '../interfaces/Params';

export const useGame = (params?: Params) => {
  return useQuery(['games', params], () => GameService.getAll(params), {
    keepPreviousData: true
  });
};
