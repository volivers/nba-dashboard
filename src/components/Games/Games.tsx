import React, { useState } from 'react';
import { useGame } from '../../hooks/useGame';
import Logo from '../Logo/Logo';
import classNames from 'classnames';
import PaginatedTable from '../Table/PaginatedTable';
import { formatDate } from '../../utils/formatDate';

export interface GamesProps {
  activeGameId?: number;
  activeTeamId: number;
  onClickGame: (game: number) => void;
}

const Games = ({ activeGameId, activeTeamId, onClickGame }: GamesProps) => {
  const [currentPage, setCurrentPage] = useState(1);
  const { data: results, isLoading } = useGame({ page: currentPage, perPage: 5, teamIds: [activeTeamId] });

  if (isLoading) return <p>Loading team games...</p>;

  return (
    <PaginatedTable
      description="Pick a game to view the stats."
      isLastPage={!results?.meta.nextPage || results?.meta.nextPage === results?.meta.currentPage}
      onPageChange={setCurrentPage}
      page={currentPage}
      title="Team Games"
    >
      <thead>
        <tr>
          <th>Game</th>
          <th>Date</th>
          <th>Score</th>
        </tr>
      </thead>
      <tbody>
        {results?.data.map(({ date, homeTeam, homeTeamScore, id, visitorTeam, visitorTeamScore }) => (
          <tr
            className={classNames({ 'table-active': activeGameId === id })}
            key={id}
            onClick={() => onClickGame(id)}
            role="button"
          >
            <td className="text-nowrap">
              <Logo name={visitorTeam.abbreviation} />
              <span className={classNames({ 'fw-bold': visitorTeamScore > homeTeamScore })}>
                {visitorTeam.name}
              </span>
              <span className="ms-2">@</span>
              <Logo name={homeTeam.abbreviation} />
              <span className={classNames({ 'fw-bold': homeTeamScore > visitorTeamScore })}>
                {homeTeam.name}
              </span>
            </td>
            <td className="text-nowrap">
              {formatDate(date)}
            </td>
            <td className="text-nowrap">
              {visitorTeamScore}-{homeTeamScore}
            </td>
          </tr>
        ))}
      </tbody>
    </PaginatedTable>
  );
};

export default Games;
