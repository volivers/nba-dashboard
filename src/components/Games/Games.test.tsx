import { render, screen } from '@testing-library/react';
import Games, { GamesProps } from './Games';
import { useGame } from '../../hooks/useGame';
import { deserializeGameEntity, mockedGameEntity } from '../../entities/GameEntity';
import { deserializePaginationEntity, mockedPaginationEntity } from '../../entities/PaginationEntity';
import { GameResponse } from '../../services/GameService';

jest.mock('../../hooks/useGame');
const mockedUseGame = useGame as jest.Mock;

describe('Games', () => {
  let props: GamesProps;
  let mockedGameResponse: GameResponse;

  beforeEach(() => {
    props = {
      activeTeamId: 2,
      onClickGame: jest.fn(),
    };

    mockedGameResponse = {
      data: [deserializeGameEntity(mockedGameEntity)],
      meta: deserializePaginationEntity(mockedPaginationEntity),
    };

    mockedUseGame.mockResolvedValue({
      ...mockedGameResponse,
      isLoading: false,
    });
  });

  it('renders properly', () => {
    const { container } = render(<Games {...props} />);

    expect(container).toMatchSnapshot();
  });

  describe('when games are loading', () => {
    beforeEach(() => {
      mockedUseGame.mockReturnValue({
        ...mockedGameResponse,
        isLoading: true,
      });
    });

    it('renders a loading message', () => {
      render(<Games {...props} />);

      expect(screen.getByText(/Loading team games.../)).toBeInTheDocument();
    });
  });
});
