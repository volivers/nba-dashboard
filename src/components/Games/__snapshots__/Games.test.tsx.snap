// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Games renders properly 1`] = `
<div>
  <div>
    <h1
      class="my-3"
    >
      Team Games
    </h1>
    <div
      class="d-flex justify-content-between"
    >
      <p>
        <small>
          Pick a game to view the stats.
        </small>
      </p>
      <ul
        class="mt-3 pagination"
      >
        <li
          class="page-item disabled"
        >
          <span
            class="page-link"
          >
            <span
              aria-hidden="true"
            >
              ‹
            </span>
            <span
              class="visually-hidden"
            >
              Previous
            </span>
          </span>
        </li>
        <li
          class="page-item disabled"
        >
          <span
            class="page-link"
          >
            <span
              aria-hidden="true"
            >
              ›
            </span>
            <span
              class="visually-hidden"
            >
              Next
            </span>
          </span>
        </li>
      </ul>
    </div>
    <div
      class="table-responsive"
    >
      <table
        class="table table-hover"
      >
        <thead>
          <tr>
            <th>
              Game
            </th>
            <th>
              Date
            </th>
            <th>
              Score
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
</div>
`;
